runs.py

```diff
@@ -120,6 +121,46 @@ def removals(self) -> list[str]:
         all_removals = self.neow_bonus.cards_removed + event_removals + store_removals
         return all_removals
 
+    @property
+    def character_streak(self) -> StreakInfo:
+        """Gets the run position in the character streak."""
+        return self._get_streak(True)
+
+    @property
+    def rotating_streak(self) -> StreakInfo:
+        """Gets the run position in the rotating streak."""
+        return self._get_streak(False)
+
+    def _get_streak(self, is_character_streak: bool) -> StreakInfo:
+        if self.won:
+            streak_total = 1
+            position_in_streak = 1
+
+            def loop_cached_runs(matched_key: str, should_iterate_position: bool):
+                nonlocal streak_total, position_in_streak
+                if is_character_streak:
+                    matched_key = matched_key + "_char"
+                if matched_key in self.matched:
+                    current_run = self.matched[matched_key]
+                    while current_run.won:
+                        # iterate the streak
+                        streak_total += 1
+                        if should_iterate_position:
+                            position_in_streak += 1
+
+                        if matched_key in current_run.matched:
+                            current_run = current_run.matched[matched_key]
+                        else:
+                            break
+            loop_cached_runs("prev", True)
+            loop_cached_runs("next", False)
+            return StreakInfo(streak_total, position_in_streak)
+        return StreakInfo(0, 0)
+
+class StreakInfo(NamedTuple):
```

Comment:
This class should also include a "is this streak currently ongoing" boolean, and the two properties should cache the result (using `self._cache`) if the streak is not currently ongoing, to reduce iterating over it for no reason.

Reply:
how should we store it in cache? I'm wondering if maybe we only need to store an array of the timestamps (filenames) of the runs. Or maybe a dict of the timestamp with its position. But seems kinda weird to do it that way if it's just going to be ordered anyways.

So we'd store something like [1663885722, 1663885723, 1663885724] and if we needed 1663885723 we'd get it's index in the list to get its position in the streak.

Also on the current run we're not going to display this right? We could say it's part of the streak but we don't know the outcome

Reply:
No, I just mean doing like `if "rotating_streak" not in self._cache: self._cache["rotating_streak"] = <blah function here>; return self._cache["rotating_streak"]`

Reply:
We already do that for other stuff to avoid re-computing it: https://github.com/Spireblight/Spireblight/blob/main/gamedata.py#L838-L860

Reply:
I don't think it's quite so simple as seed. We could just store the rotating_streak in cache as a number and such, but the problem comes when we're displaying which run this currently is in the streak (ex: Rotating 5/7 for the 5th win of the streak).

If we're not caching each run's position as well, we'll have to loop over the runs around it to calculate it (at least the `prev` keys). So we can cache the ongoing streak as an int in cache, but I think we'll also have to cache either a dict[timestamp, position] or just a list[timestamp] and get the positions by index.

Hopefully this makes sense.

Reply:
I meant to cache the `StreakInfo` class, which has the position built in.

Reply:
the StreakInfo class is contextual to whichever run we're viewing at the time. But we are able to view multiple runs at a time via the Run History. So we would still need multiple StreakInfo classes to store for each run we're viewing.

Reply:
Sure, but the `_cache` dict is exclusive to any one given run object, meaning we can store that run's streak info in there, without worrying about the other runs (which will cache their own).

Reply:
I see where our misunderstanding is. The cache in FileParser is just a field on FileParser, rather than an overarching cache like I'm imagining. I find it odd to use the term cache this way.

Reply:
I made this comment before seeing your latest

Reply:
Oh yeah, I probably could have named that attribute better haha. oh well!

Reply:
Follow up question, do we want this to display for the Save file (current run)? I opted not to, but we certainly could. Right now it's only showing for Victories, we could hide it for losses and show for victories/ongoing run

Reply:
I think showing an ongoing streak for the current run could be nice, I'll leave that up to you. Definitely hide it for losses.